docs/development/faq.md

```diff
@@ -0,0 +1,17 @@
+# Developer FAQ
+
+## What is an `outlier`?
+
+An `outlier` is an arbitrary floating event in the DAG (as opposed to being
+inline with the current DAG). It also means that we don't have the state events
+backfilled on the homeserver and we trust the events *claimed* auth events rather
+than those we calculate and verify to be correct. 
+
+An event can be unmarked as an `outlier` once we fetch all of its `prev_events` (you will see some `ex_outlier` code around this).
```

Comment:
I think it's slightly different. The thing that makes an `outlier` an outlier is that we haven't figured out the state for the room at that point in the DAG. 

We won't *necessarily* have its prev_events in the database, but it's entirely possible that we *will*. Conversely, it's possible for us not to have all the prev_events for non-outlier event (in which case, that event would be a "backward extremity").

The process of transforming an outlier to a non-outlier involves figuring out the room state at that point in the DAG - *either* by having the state at all the `prev_events` and resolving the state between them, or simply by asking another server.

auth events are a whole extra level of complexity, but even for an outlier we should have all the events in the auth chain. https://github.com/matrix-org/synapse/blob/cd5fcd2731182aa48aa7c0a44c7e547681021296/synapse/handlers/federation.py#L2196-L2199 is quite a specific bit of code relating to one part of processing incoming events (and to be honest, I'm not convinced it does the right thing).  I would leave auth events out of this explanation, tbh.

Reply:
Thanks for all info @richvdh 😀

These seem so closely coupled and aren't mutually exclusive in real life situations so it's hard for me to make a clear description 🤯

I've updated with something more inline with your description but feel free to rewrite.

---

I feel like I don't understand where an `outlier` becomes an `ex_outlier`.

Say we fetch a missing auth event for an event, we mark that auth event as an `outlier`. Then at what point later, do we get the event again as a non-outlier? From your statement, it's not when we have fetched all of the `prev_events` as those could already be all persisted in the db.

There is [`_update_outliers_txn`](https://github.com/matrix-org/synapse/blob/858363d0b7e58fd71875b25d183537bb3b5a397f/synapse/storage/databases/main/events.py#L1225) which handles `ex_outlier` stuff but it's confusing where/when we will come across the same `outlier` event again while persisting.

Based on the DAG below, as a remote federating server, my assumption is that we process event `D` which fetches the `state_event` as an `outlier`. Then as we backfill more, and process `state_event` directly, it becomes an `ex_outlier`.

[Mermaid live editor playground link](https://mermaid-js.github.io/mermaid-live-editor/edit/#eyJjb2RlIjoiZmxvd2NoYXJ0IEJUXG4gICAgc3ViZ3JhcGggbGl2ZVxuICAgICAgICBEIC0tLS0-IEMgLS0tLT4gQiAtLS0tPiBzdGF0ZV9ldmVudCAtLS0tPiBBXG4gICAgZW5kXG4gICAgXG5cbiAgICBEIC0uLT4gc3RhdGVfZXZlbnRcblxuICAgIFxuICAgICUlIGFsaWdubWVudCBsaW5rcyB0byBtYWtlIHRoZSBncmFwaCBwb2ludCB1cHdhcmRcbiAgICBudWxsWyBdIC0tLSBBXG4gICAgJSUgbWFrZSB0aGUgbGlua3MgYW5kIG5vZGVzIGludmlzaWJsZVxuICAgIHN0eWxlIG51bGwgZGlzcGxheTpub25lO1xuICAgIGxpbmtTdHlsZSA1IHN0cm9rZS13aWR0aDoycHgsZmlsbDpub25lLHN0cm9rZTpub25lO1xuXG4gICAgIiwibWVybWFpZCI6Int9IiwidXBkYXRlRWRpdG9yIjpmYWxzZSwiYXV0b1N5bmMiOnRydWUsInVwZGF0ZURpYWdyYW0iOmZhbHNlfQ)

![](https://user-images.githubusercontent.com/558581/127402622-c1f94787-626d-423c-b597-a183eb5b9b0a.png)


I'm still a bit wary on a situation where we could have all of the `prev_events` for the `state_event` but still mark it as an `outlier`. Perhaps with gaps in the DAG and the event hits it just right 🤷

Reply:
> I've mainly been working with floating outliers so I don't think I had the complete picture.

I think I might have misled you here. From Synapse's point of view, all outliers are considered to be floating, and in the majority of cases we indeed won't have their prev events. However, it's possible for the stars to align in a way that we do actually have their prev events in the database - at least as outliers themselves.

To take a simple example:

Two of the auth events for a `join` event in an invite-only room are the `join_rules` event and an invite event for the joiner. So let's suppose we fetch both of those events as outliers. Let's also suppose that the invite was issued immediately after changing the join rules - so the `join_rules` is the only prev-event of the `invite` event.

```
join_rules <-  invite
   ^             ^
   ............................. join
```

In that case, it so happens that we have all the prev-events of the invite event - but we're still considering it an outlier.

So I don't want you to imagine there is a distinction between "floating outliers" and "regular outliers" - they are all just outliers in terms of how we handle them in Synapse.

> Say we fetch a missing auth event for an event, we mark that auth event as an outlier. Then at what point later, do we get the event again as a non-outlier? From your statement, it's not when we have fetched all of the prev_events as those could already be all persisted in the db.

right. It becomes a non-outlier in the situation where we process it as a regular event as part of the timeline - normally via backfill.

> Based on the DAG below, as a remote federating server, my assumption is that we process event D which fetches the state_event as an outlier. Then as we backfill more, and process state_event directly, it becomes an ex_outlier.

exactly so, yes.

> I'm still a bit wary on a situation where we could have all of the `prev_events` for the `state_event` but still mark it as an `outlier`. Perhaps with gaps in the DAG and the event hits it just right :shrug:

yup. Imagine there is another fork in the DAG pointing to A:

[![](https://mermaid.ink/img/eyJjb2RlIjoiZmxvd2NoYXJ0IEJUXG4gICAgJSUgYWxpZ25tZW50IGxpbmtzIHRvIG1ha2UgdGhlIGdyYXBoIHBvaW50IHVwd2FyZFxuICAgIG51bGxbIF0gLS0tIEFcblxuICAgIHN1YmdyYXBoIGxpdmVcbiAgICAgICAgRCAtLS0tPiBDIC0tLS0-IEIgLS0tLT4gc3RhdGVfZXZlbnQgLS0tLT4gQVxuICAgICAgICBaIC0tPiBZIC0tLS0tPiBBXG4gICAgZW5kXG4gICAgXG5cbiAgICBEIC0uLT4gc3RhdGVfZXZlbnRcblxuICAgIFxuXG4gICAgJSUgbWFrZSB0aGUgbGlua3MgYW5kIG5vZGVzIGludmlzaWJsZVxuICAgIHN0eWxlIG51bGwgZGlzcGxheTpub25lO1xuICAgIGxpbmtTdHlsZSAwIHN0cm9rZS13aWR0aDoycHgsZmlsbDpub25lLHN0cm9rZTpub25lO1xuXG4gICAgIiwibWVybWFpZCI6e30sInVwZGF0ZUVkaXRvciI6ZmFsc2UsImF1dG9TeW5jIjp0cnVlLCJ1cGRhdGVEaWFncmFtIjpmYWxzZX0)](https://mermaid-js.github.io/mermaid-live-editor/edit/##eyJjb2RlIjoiZmxvd2NoYXJ0IEJUXG4gICAgJSUgYWxpZ25tZW50IGxpbmtzIHRvIG1ha2UgdGhlIGdyYXBoIHBvaW50IHVwd2FyZFxuICAgIG51bGxbIF0gLS0tIEFcblxuICAgIHN1YmdyYXBoIGxpdmVcbiAgICAgICAgRCAtLS0tPiBDIC0tLS0-IEIgLS0tLT4gc3RhdGVfZXZlbnQgLS0tLT4gQVxuICAgICAgICBaIC0tPiBZIC0tLS0-IEFcbiAgICBlbmRcbiAgICBcblxuICAgIEQgLS4tPiBzdGF0ZV9ldmVudFxuXG4gICAgXG5cbiAgICAlJSBtYWtlIHRoZSBsaW5rcyBhbmQgbm9kZXMgaW52aXNpYmxlXG4gICAgc3R5bGUgbnVsbCBkaXNwbGF5Om5vbmU7XG4gICAgbGlua1N0eWxlIDAgc3Ryb2tlLXdpZHRoOjJweCxmaWxsOm5vbmUsc3Ryb2tlOm5vbmU7XG5cbiAgICAiLCJtZXJtYWlkIjoie30iLCJ1cGRhdGVFZGl0b3IiOmZhbHNlLCJhdXRvU3luYyI6dHJ1ZSwidXBkYXRlRGlhZ3JhbSI6ZmFsc2V9)

When we receive Y, we'll backfill A. So then we have `state_event`'s `prev_events` - but `state_event` is still an outlier at this point.

Reply:
Thanks for the clarification here. I've added a paragraph explaining that there is no distinction.

I like these examples and edge cases. I'm tempted to add them into the docs but I think I'll save that for another iteration.